### ISSUE TITLE
deis run: 'utf-8' codec can't decode byte 0xe7 in position 0: invalid continuation byte

### ISSUE THREAD

**Issue by felixbuenemann** [deis/controller#862]
I'm getting the following error on 2.1.0 when trying to do a `deis run /bin/bash` (command doesn't matter):

```
Running '/bin/bash'...
Error: Unknown Error (503): {"detail":"shop-backend-v7-run-ivbxw (run): 'utf-8' codec can't decode byte 0xe7 in position 0: invalid continuation byte"}
```

The logs for the deis-controller pod show:

```
INFO [shop-backend]: buenemann on shop-backend-v7-run-ivbxw runs ''/bin/bash''
INFO:api.models.app:[shop-backend]: buenemann on shop-backend-v7-run-ivbxw runs ''/bin/bash''
INFO run shop-backend-v7-run-ivbxw, img home/shop-backend:git-5cb6a5fe/push/slug.tgz, entrypoint /runner/init, cmd "'/bin/bash'"
INFO:scheduler:run shop-backend-v7-run-ivbxw, img home/shop-backend:git-5cb6a5fe/push/slug.tgz, entrypoint /runner/init, cmd "'/bin/bash'"
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): 10.3.0.1
ERROR:root:'utf-8' codec can't decode byte 0xe7 in position 0: invalid continuation byte
Traceback (most recent call last):
  File "/app/api/models/app.py", line 631, in run
    **kwargs
  File "/app/scheduler/__init__.py", line 377, in run
    manifest = self._build_pod_manifest(namespace, name, image, **kwargs)
  File "/app/scheduler/__init__.py", line 348, in _build_pod_manifest
    self._set_container(namespace, container_name, container, **kwargs)
  File "/app/scheduler/__init__.py", line 461, in _set_container
    self.get_secret(namespace, secret_name)
  File "/app/scheduler/__init__.py", line 1031, in get_secret
    secrets['data'][key] = value.decode(encoding='UTF-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe7 in position 0: invalid continuation byte
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/rest_framework/views.py", line 463, in dispatch
    response = handler(request, *args, **kwargs)
  File "/app/api/views.py", line 225, in run
    rc, output = app.run(self.request.user, request.data['command'])
  File "/app/api/models/app.py", line 637, in run
    raise ServiceUnavailable(err) from e
api.exceptions.ServiceUnavailable: shop-backend-v7-run-ivbxw (run): 'utf-8' codec can't decode byte 0xe7 in position 0: invalid continuation byte
10.2.21.6 "POST /v2/apps/shop-backend/run HTTP/1.1" 503 123 "Deis Client v2.1.0"
```

I'm running deis/workflow-v2.1.0 on a kube-aws 1.2.4 cluster.

**Comment by felixbuenemann** [comment]
Ahh, it's trying to decode a value that's stored unencoded in the app's secrets:

```ruby
apiVersion: v1
data:
  builder-storage: czM=
  deis-app: c2hvcC1iYWNrZW5k
  deis-minio-service-host: Tm9uZQ==
  deis-minio-service-port: Tm9uZQ==
  port: "5000"
  rails-env: cHJvZHVjdGlvbg==
  slug-url: aG9tZS9zaG9wLWJhY2tlbmQ6Z2l0LTVjYjZhNWZlL3B1c2gvc2x1Zy50Z3o=
  workflow-release: djc=
kind: Secret
metadata:
  creationTimestamp: 2016-07-10T20:02:36Z
  labels:
    app: shop-backend
    heritage: deis
    type: env
    version: v7
  name: shop-backend-v7-env
  namespace: shop-backend
  resourceVersion: "103917"
  selfLink: /api/v1/namespaces/shop-backend/secrets/shop-backend-v7-env
  uid: 3f052b36-46d9-11e6-84aa-0274740f37d1
type: Opaque
```

The interesting bit is `port: "5000"` which is NOT base64 encoded, but if you try to base64 decode it anyways you get `"\xE7M4"` which leads to the above error `can't decode byte 0xe7 in position 0`.

**Comment by helgi** [comment]
Ah yeah, seen that problem before and I fixed it (I thought) I need to track down if it made it into 2.1 or if it is slatted for 2.2

**Comment by felixbuenemann** [comment]
Great!

For now I worked around the issue by doing a `kubectl edit --namespace=shop-backend secrets/shop-backend-v7-env` and replacing the value for `port` with `"NTAwMA=="`.

**Comment by helgi** [comment]
Awesome :) One thing that may happen is it gets reverted to port number since the secret for env var is per release.

The peculiar behaviour is that this only ever happened to `PORT`, from what I could tell at the time

**Comment by felixbuenemann** [comment]
… which worked for exactly one invocation, after that the secret was reverted to the old value :-(

**Comment by felixbuenemann** [comment]
OK, I used `deis config:set PORT=5000` which fixed it. So this has probably to do with the default value set by the buildpack.

**Comment by helgi** [comment]
The fix I was thinking about supposedly fixed it is in 2.1 :( https://github.com/deis/controller/commit/2f27dd7fc40e27c3ea5838db06da216f46cb520b#diff-3b6e1f09121075ad67f216aa2bbfc9a6L1168 

Thanks for finding the workaround. I'll try to tackle this for 2.2!

**Comment by helgi** [comment]
Going to reopen this again, for now

**Comment by helgi** [comment]
@felixbuenemann Are you able to run `workflow-dev` to test if this has been fixed?

I wrote e2e tests that do a `git push` and then `deis run env` and I am not seeing any problems right now https://github.com/deis/workflow-e2e/pull/252

**Comment by felixbuenemann** [comment]
I didn't spin up a whole workflow-dev environment, but I can confirm that the issue is fixed, after updating my controller on v2.1.0 to the `quay.io/deisci/controller:git-abdf9cb` image (that's the build on the merge commit for PR #863).

I created a fresh app using the [heroku/ruby-getting-started](https://github.com/heroku/ruby-getting-started.git) example app, deployed it and ran `deis run -- rake -T` successfully.
I also inspected the environment secrets using `kubectl` and confirmed that the default value for the `PORT` environment variable is now properly base64 encoded as `NTAwMA==`.
